src/command/attributecommand.js

```diff
@@ -0,0 +1,201 @@
+		const schema = this.editor.document.schema;
+
+		if ( selection.isCollapsed ) {
+			// Check whether schema allows for a test with `attributeKey` in caret position.
+			return schema.checkAtPosition( selection.getFirstPosition(), '$text', this.attributeKey );
+		} else {
+			const ranges = selection.getRanges();
+
+			// For all ranges, check nodes in them until you find a node that is allowed to have `attributeKey` attribute.
+			for ( let range of ranges ) {
+				const walker = new TreeWalker( { boundaries: range, mergeCharacters: true } );
+				let last = walker.position;
+				let step = walker.next();
+
+				// Walk the range.
+				while ( !step.done ) {
+					// If returned item does not have name property, it is a treeModel.TextFragment.
+					const name = step.value.item.name || '$text';
+
+					if ( schema.checkAtPosition( last, name, this.attributeKey ) ) {
+						// If we found a node that is allowed to have the attribute, return true.
+						return true;
+					}
+
+					last = walker.position;
+					step = walker.next();
+				}
+			}
+		}
+
+		// If we haven't found such node, return false.
+		return false;
+	}
+
+	/**
+	 * Executes the command: adds or removes attributes to nodes or selection.
+	 *
+	 * If the command is active (`value == true`), it will remove attributes. Otherwise, it will set attributes.
+	 *
+	 * The execution result differs, depending on the {@link core.treeModel.Document#selection}:
+	 * * if selection is on a range, the command applies the attribute on all nodes in that ranges
+	 * (if they are allowed to have this attribute by the{@link core.treeModel.Schema schema}),
+	 * * if selection is collapsed in non-empty node, the command applies attribute to the {@link core.treeModel.Document#selection}
+	 * itself (note that typed characters copy attributes from selection),
+	 * * if selection is collapsed in empty node, the command applies attribute to the parent node of selection (note
+	 * that selection inherits all attributes from a node if it is in empty node).
+	 *
+	 * If the command is disabled (`isEnabled == false`) when it is executed, nothing will happen.
+	 *
+	 * @private
+	 * @param {Boolean} [forceValue] If set it will force command behavior. If `true`, command will apply attribute,
+	 * otherwise command will remove attribute. If not set, command will look for it's current value to decide what it should do.
+	 */
+	_execute( forceValue ) {
+		const document = this.editor.document;
+		const selection = document.selection;
+		const value = ( forceValue === undefined ) ? !this.value : forceValue;
+
+		if ( selection.isCollapsed ) {
+			let selectionParent = selection.getFirstPosition().parent;
+
+			if ( selectionParent.getChildCount() === 0 ) {
+				// If selection is collapsed and in empty node, operate on stored selection attributes.
+				const storeKey = Selection.getStoreAttributeKey( this.attributeKey );
+
+				document.enqueueChanges( () => {
```

Comment:
I wouldn't go this way. If we require using `enqueueChanges()` then let's make it clear that you should call it before you do something. Otherwise we risk creating an unclear situation in which the developers sometimes will have to call it and sometimes it will be called automatically.

PS. I'd even consider throwing if you're trying to make some changes when not inside `enqueueChanges()`.